module/apmhttp/clienttrace.go

```diff
@@ -0,0 +1,97 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apmhttp
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http/httptrace"
+
+	"go.elastic.co/apm"
+)
+
+// WithClientTrace returns a ClientOption for
+// tracing events within HTTP client requests.
+func WithClientTrace() ClientOption {
+	return func(rt *roundTripper) {
+		rt.traceRequests = true
+	}
+}
+
+type connectKey struct {
+	network, addr string
+}
+
+type requestTracer struct {
+	DNS,
+	TLS,
+	Request,
+	Response *apm.Span
+
+	Connects map[connectKey]*apm.Span
+}
+
+func withClientTrace(ctx context.Context, tx *apm.Transaction, parent *apm.Span) (context.Context, *requestTracer) {
+	r := requestTracer{
+		Connects: make(map[connectKey]*apm.Span),
+	}
+
+	return httptrace.WithClientTrace(ctx, &httptrace.ClientTrace{
+		DNSStart: func(i httptrace.DNSStartInfo) {
+			r.DNS = tx.StartSpan(fmt.Sprintf("DNS %s", i.Host), "http.dns", parent)
+		},
+
+		DNSDone: func(i httptrace.DNSDoneInfo) {
+			r.DNS.End()
+		},
+
+		ConnectStart: func(network, addr string) {
+			span := tx.StartSpan(fmt.Sprintf("Connect %s", addr), "http.connect", parent)
+			r.Connects[connectKey{network: network, addr: addr}] = span
+		},
+
+		ConnectDone: func(network, addr string, err error) {
+			span := r.Connects[connectKey{network: network, addr: addr}]
+			span.End()
+		},
+
+		TLSHandshakeStart: func() {
+			r.TLS = tx.StartSpan("TLS", "http.tls", parent)
+		},
+
+		TLSHandshakeDone: func(_ tls.ConnectionState, _ error) {
+			r.TLS.End()
+		},
+
+		WroteRequest: func(info httptrace.WroteRequestInfo) {
+			r.Request = tx.StartSpan("Request", "http.request", parent)
```

Comment:
Isn't this when we want to _end_ the request span? I think we want to start it with GotConn

Reply:
You're right, this was not translated properly when converting it from `Server Processing`.  Looking at this again and seeing:

```
	// WroteRequest is called with the result of writing the
	// request and any body. It may be called multiple times
	// in the case of retried requests.
```

I think we should make multiple request spans in case of retries, so if `r.Request != nil r.Request.End()` - objections?

Reply:
Cancel that, will test what happens in this case, not clear if gotconn is called again too

Reply:
Now that the request span doesn't cover the server response time I'm missing that span representing the server time

<img width="243" alt="image" src="https://user-images.githubusercontent.com/83483/89247636-c6516a80-d5db-11ea-97ef-347835fb4f8d.png">

I follow that a distributed trace into your own systems would show up there but this makes it non-trivial to monitor time in external services - any ideas here?  Maybe an option for `WithClientTrace` to create spans for that external time?  Is this not actually an issue?

Reply:
> I follow that a distributed trace into your own systems would show up there but this makes it non-trivial to monitor time in external services - any ideas here?

Fair point, I was really thinking about wholly-owned distributed systems.

Going back to https://w3c.github.io/resource-timing/#processing-model, "Request" is actually `responseStart - requestStart`. How about we do the same? `GotFirstResponseByte - GotConn`? If it's really interesting to know when the client finished writing the request, we could add "marks" like the RUM agent does.